osu.Game/Online/OnlineStatusNotifier.cs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Extensions.ObjectExtensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Sprites;
+using osu.Framework.Screens;
+using osu.Game.Online.API;
+using osu.Game.Online.Multiplayer;
+using osu.Game.Online.Spectator;
+using osu.Game.Overlays;
+using osu.Game.Overlays.Notifications;
+using osu.Game.Screens.OnlinePlay;
+
+namespace osu.Game.Online
+{
+    public partial class OnlineStatusNotifier : Component
+    {
+        private readonly Func<IScreen> getCurrentScreen;
+
+        [Resolved]
+        private MultiplayerClient multiplayerClient { get; set; } = null!;
+
+        [Resolved]
+        private SpectatorClient spectatorClient { get; set; } = null!;
+
+        [Resolved]
+        private INotificationOverlay? notificationOverlay { get; set; }
+
+        private IBindable<APIState> apiState = null!;
+        private IBindable<bool> multiplayerState = null!;
+        private IBindable<bool> spectatorState = null!;
+        private bool forcedDisconnection;
+
+        public OnlineStatusNotifier(Func<IScreen> getCurrentScreen)
```

Comment:
I rewrote this class with a hope of simplifying things but I'm not sure if what I came up with is better or worse:

```diff
diff --git a/osu.Game/Online/IStatefulUserHubClient.cs b/osu.Game/Online/IStatefulUserHubClient.cs
index 86105dd629..ec35be8ff7 100644
--- a/osu.Game/Online/IStatefulUserHubClient.cs
+++ b/osu.Game/Online/IStatefulUserHubClient.cs
@@ -1,6 +1,7 @@
 // Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
 // See the LICENCE file in the repository root for full licence text.
 
+using System;
 using System.Threading.Tasks;
 
 namespace osu.Game.Online
@@ -13,6 +14,11 @@ namespace osu.Game.Online
     /// </summary>
     public interface IStatefulUserHubClient
     {
+        /// <summary>
+        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
+        /// </summary>
+        public event Action? Disconnecting;
+
         Task DisconnectRequested();
     }
 }
diff --git a/osu.Game/Online/Multiplayer/MultiplayerClient.cs b/osu.Game/Online/Multiplayer/MultiplayerClient.cs
index 79f46c2095..4d473a4b60 100644
--- a/osu.Game/Online/Multiplayer/MultiplayerClient.cs
+++ b/osu.Game/Online/Multiplayer/MultiplayerClient.cs
@@ -32,6 +32,8 @@ public abstract partial class MultiplayerClient : Component, IMultiplayerClient,
 
         public Action<Room, string>? PresentMatch { protected get; set; }
 
+        public event Action? Disconnecting;
+
         /// <summary>
         /// Invoked when any change occurs to the multiplayer room.
         /// </summary>
@@ -87,11 +89,6 @@ public abstract partial class MultiplayerClient : Component, IMultiplayerClient,
         /// </summary>
         public event Action? ResultsReady;
 
-        /// <summary>
-        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
-        /// </summary>
-        public event Action? Disconnecting;
-
         /// <summary>
         /// Whether the <see cref="MultiplayerClient"/> is currently connected.
         /// This is NOT thread safe and usage should be scheduled.
diff --git a/osu.Game/Online/OnlineStatusNotifier.cs b/osu.Game/Online/OnlineStatusNotifier.cs
index 0cf672ac3c..d0a6b83681 100644
--- a/osu.Game/Online/OnlineStatusNotifier.cs
+++ b/osu.Game/Online/OnlineStatusNotifier.cs
@@ -17,6 +17,9 @@
 
 namespace osu.Game.Online
 {
+    /// <summary>
+    /// Handles various scenarios where connection is lost and we need to let the user know what and why.
+    /// </summary>
     public partial class OnlineStatusNotifier : Component
     {
         private readonly Func<IScreen> getCurrentScreen;
@@ -33,7 +36,11 @@ public partial class OnlineStatusNotifier : Component
         private IBindable<APIState> apiState = null!;
         private IBindable<bool> multiplayerState = null!;
         private IBindable<bool> spectatorState = null!;
-        private bool forcedDisconnection;
+
+        /// <summary>
+        /// This flag will be set to <c>true</c> when the user has been notified so we don't show more than one notification.
+        /// </summary>
+        private bool userNotified;
 
         public OnlineStatusNotifier(Func<IScreen> getCurrentScreen)
         {
@@ -51,59 +58,68 @@ private void load(IAPIProvider api)
             spectatorClient.Disconnecting += notifyAboutForcedDisconnection;
         }
 
-        private void notifyAboutForcedDisconnection()
-        {
-            if (forcedDisconnection)
-                return;
-
-            forcedDisconnection = true;
-            notificationOverlay?.Post(new SimpleErrorNotification
-            {
-                Icon = FontAwesome.Solid.ExclamationCircle,
-                Text = "You have been logged out on this device due to a login to your account on another device."
-            });
-        }
-
         protected override void LoadComplete()
         {
             base.LoadComplete();
 
-            apiState.BindValueChanged(_ =>
+            apiState.BindValueChanged(state =>
             {
-                if (apiState.Value == APIState.Online)
-                    forcedDisconnection = false;
+                if (state.NewValue == APIState.Online)
+                {
+                    userNotified = false;
+                    return;
+                }
 
-                Scheduler.AddOnce(updateState);
+                if (state.NewValue == APIState.Offline)
+                {
+                    if (getCurrentScreen() is OnlinePlayScreen)
+                    {
+                        userNotified = true;
+                        notificationOverlay?.Post(new SimpleErrorNotification
+                        {
+                            Icon = FontAwesome.Solid.ExclamationCircle,
+                            Text = "Connection to API was lost. Can't continue with online play."
+                        });
+                    }
+                }
             });
-            multiplayerState.BindValueChanged(_ => Scheduler.AddOnce(updateState));
-            spectatorState.BindValueChanged(_ => Scheduler.AddOnce(updateState));
-        }
 
-        private void updateState()
-        {
-            if (forcedDisconnection)
-                return;
-
-            if (apiState.Value == APIState.Offline && getCurrentScreen() is OnlinePlayScreen)
+            multiplayerState.BindValueChanged(connected =>
             {
-                notificationOverlay?.Post(new SimpleErrorNotification
+                if (connected.NewValue)
                 {
-                    Icon = FontAwesome.Solid.ExclamationCircle,
-                    Text = "API connection was lost. Can't continue with online play."
-                });
-                return;
-            }
+                    userNotified = false;
+                    return;
+                }
 
-            if (!multiplayerClient.IsConnected.Value && multiplayerClient.Room != null)
+                if (multiplayerClient.Room != null)
+                {
+                    userNotified = true;
+                    notificationOverlay?.Post(new SimpleErrorNotification
+                    {
+                        Icon = FontAwesome.Solid.ExclamationCircle,
+                        Text = "Connection to the multiplayer server was lost. Exiting multiplayer."
+                    });
+                }
+            });
+
+            spectatorState.BindValueChanged(_ =>
             {
+                // TODO: handle spectator server failure somehow?
+            });
+        }
+
+        private void notifyAboutForcedDisconnection()
+        {
+            if (!userNotified)
+            {
+                userNotified = true;
                 notificationOverlay?.Post(new SimpleErrorNotification
                 {
                     Icon = FontAwesome.Solid.ExclamationCircle,
-                    Text = "Connection to the multiplayer server was lost. Exiting multiplayer."
+                    Text = "You have been logged out on this device due to a login to your account on another device."
                 });
             }
-
-            // TODO: handle spectator server failure somehow?
         }
 
         protected override void Dispose(bool isDisposing)
diff --git a/osu.Game/Online/Spectator/SpectatorClient.cs b/osu.Game/Online/Spectator/SpectatorClient.cs
index ca4ec52f4a..0cc78734df 100644
--- a/osu.Game/Online/Spectator/SpectatorClient.cs
+++ b/osu.Game/Online/Spectator/SpectatorClient.cs
@@ -70,9 +70,6 @@ public abstract partial class SpectatorClient : Component, ISpectatorClient
         /// </summary>
         public virtual event Action<int, long>? OnUserScoreProcessed;
 
-        /// <summary>
-        /// Invoked just prior to disconnection requested by the server via <see cref="IStatefulUserHubClient.DisconnectRequested"/>.
-        /// </summary>
         public event Action? Disconnecting;
 
         /// <summary>

```

Reply:
I'm not sure either, so I'm going to test this and apply it as per adage of "the reviewer is always right" if I see no issues with this.

Reply:
The patch needed further changes (not enough scheduling, as `IsConnected` changes are not update thread, also not enough checking of `userNotified`), but I made it work. See how you like the result.